File: content/chapters/compute/lecture/demo/create-thread/create_thread.py
#!/usr/bin/env python

# SPDX-License-Identifier: BSD-3-Clause

from os import getpid, getppid
from threading import Thread
from time import sleep


def thread(msg):
    print(f"[thread] PID = {getpid()}; PPID = {getppid()}")
    print(f"[thread] Message from main thread: {msg}")

    sleep(5)


def main():
    print(f"[main] PID = {getpid()}; PPID = {getppid()}")

    # Create a Thread object.
    t = Thread(target=thread, args=("OS Rullz!",))
    # Start the thread.
    t.start()

    # Wait until thread finishes.
    # We can't kill a thread. Why?
    t.join()


if __name__ == "__main__":
    main()

File: content/compute/lecture/demo/create-thread/create_thread.c
// SPDX-License-Identifier: BSD-3-Clause

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <pthread.h>

#include "utils/utils.h"

void *thread_func(void *arg)
{
	char *msg = arg;

	printf("[thread] PID = %d; PPID = %d\n", getpid(), getppid());
	printf("[thread] Message from main thread: %s\n", msg);

	sleep(5);

	return NULL;
}

int main(void)
{
	pthread_t tid;
	int rc;
	char *msg = "SO Rullz!";

	printf("[main] PID = %d; PPID = %d\n", getpid(), getppid());

	/* Create and start a new thread. */
	rc = pthread_create(&tid, NULL, thread_func, msg);
	DIE(rc, "pthread_create");

	rc = pthread_join(tid, NULL);
	DIE(rc, "pthread_join");

	return 0;
}
